scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,386 @@
+        """
+        reconstruction_losses = []
+        kl_divergences = []
+        losses = []
+        total_batch_size = 0
+        for (i, (sample_batch, l_mean, l_var, batch_index, labels, *_)) in enumerate(
+            tensors
+        ):
+            reconstruction_loss, kl_divergence = self.model(
+                sample_batch, l_mean, l_var, batch_index, mode=i
+            )
+            loss = torch.mean(
+                reconstruction_loss + self.kl_weight * kl_divergence
+            ) * sample_batch.size(0)
+            total_batch_size += sample_batch.size(0)
+            losses.append(loss)
+            if return_details:
+                reconstruction_losses.append(reconstruction_loss.mean())
+                kl_divergences.append(kl_divergence.mean())
+
+        if return_details:
+            return reconstruction_losses, kl_divergences
+
+        averaged_loss = torch.stack(losses).sum() / total_batch_size
+        return averaged_loss
+
+    def get_discriminator_confusion(self) -> np.ndarray:
+        confusion = []
+        for i, posterior in enumerate(self.all_dataset):
+            data = torch.from_numpy(posterior.gene_dataset.X)
+            if self.use_cuda:
+                data.cuda()
+
+            z = self.model.sample_from_posterior_z(data, mode=i, deterministic=True)
+            cls_z = nn.Softmax(dim=1)(self.discriminator(z)).detach()
+
+            if self.use_cuda:
+                cls_z = cls_z.cpu()
+            cls_z = cls_z.numpy()
+
+            row = cls_z.mean(axis=0)
+            confusion.append(row)
+        return np.array(confusion)
+
+    def get_loss_magnitude(
+        self, one_sample: bool = False
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        total_reconstruction = np.zeros(self.n_dataset)
+        total_kl_divergence = np.zeros(self.n_dataset)
+        total_discriminator = np.zeros(self.n_dataset)
+
+        for tensors_list in self.data_loaders_loop():
+            reconstruction_losses, kl_divergences = self.loss(
+                tensors_list, return_details=True
+            )
+
+            discriminator_losses = self.loss_discriminator(
+                [
+                    self.model.sample_from_posterior_z(
+                        data, mode=i, deterministic=False
+                    )
+                    for (i, (data, *_)) in enumerate(tensors_list)
+                ],
+                return_details=True,
+            )
+
+            for i in range(self.n_dataset):
+                total_reconstruction[i] += reconstruction_losses[i].item()
+                total_kl_divergence[i] += kl_divergences[i].item()
+                total_discriminator[i] += discriminator_losses[i].item()
+            if one_sample:
+                break
+
+        return total_reconstruction, total_kl_divergence, total_discriminator
+
+    def get_latent(self, deterministic: bool = True) -> List[np.ndarray]:
```

Comment:
Same for this ? Posterior and not trainer ?

Reply:
It seems difficult to add this in Posterior as for now Posterior is bond to a specific GeneExpressionDataset and this function returns the latent space for all the datasets.

Reply:
cf. our conversation, got it